src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+        hashsetPauseRehashing(s);
+
+        /* Compute bucket index from bucket pointer. */
+        void *b0 = &s->tables[table_index][0];
+        size_t bucket_index = ((uintptr_t)b - (uintptr_t)b0) / sizeof(bucket);
+        assert(&s->tables[table_index][bucket_index] == b);
+
+        /* Encode position as pointer. */
+        *position = encodePositionInTable(bucket_index, pos_in_bucket, table_index);
+        return &b->elements[pos_in_bucket];
+    } else {
+        return NULL;
+    }
+}
+
+/* Clears the position of the element in the hashset and resumes rehashing. The
+ * element destructor is NOT called. The position is an opaque representation of
+ * its position as found using hashsetTwoPhasePopFindRef(). */
+void hashsetTwoPhasePopDelete(hashset *s, void *position) {
+    /* Decode position. */
+    size_t bucket_index;
+    int table_index, pos_in_bucket;
+    decodePositionInTable(position, &bucket_index, &pos_in_bucket, &table_index);
+
+    /* Delete the element and resume rehashing. */
+    bucket *b = &s->tables[table_index][bucket_index];
+    assert(b->presence & (1 << pos_in_bucket));
+    b->presence &= ~(1 << pos_in_bucket);
+    s->used[table_index]--;
+    hashsetShrinkIfNeeded(s);
+    hashsetResumeRehashing(s);
+}
+
+/* --- Scan --- */
+
+/* Scan is a stateless iterator. It works with a cursor that is returned to the
+ * caller and which should be provided to the next call to continue scanning.
+ * The hash table can be modified in any way between two scan calls. The scan
+ * still continues iterating where it was.
+ *
+ * A full scan is performed like this: Start with a cursor of 0. The scan
+ * callback is invoked for each element scanned and a new cursor is returned.
+ * Next time, call this function with the new cursor. Continue until the
+ * function returns 0.
+ *
+ * We say that an element is *emitted* when it's passed to the scan callback.
+ *
+ * Scan guarantees:
+ *
+ * - An element that is present in the hash table during an entire full scan
+ *   will be returned (emitted) at least once. (Most of the time exactly once,
+ *   but sometimes twice.)
+ *
+ * - An element that is inserted or deleted during a full scan may or may not be
+ *   returned during the scan.
+ *
+ * The hash table uses a variant of linear probing with a cursor increment
+ * rather than a regular increment of the index when probing. The scan algorithm
+ * needs to continue scanning as long as a bucket in either of the tables has
+ * ever been full. This means that we may wrap around cursor zero and still
+ * continue until we find a bucket where we can stop, so some elements can be
+ * returned twice (in the first and the last scan calls) due to this.
+ *
+ * The 'flags' argument can be used to tweak the behaviour. It's a bitwise-or
+ * (zero means no flags) of the following:
+ *
+ * - HASHSET_SCAN_EMIT_REF: Emit a pointer to the element's location in the
+ *   table is passed to the scan function instead of the actual element. This
```

Comment:
```suggestion
 *   table to the scan function instead of the actual element. This
```
This seems more right?